sstables/sstable_set.cc

```diff
@@ -464,6 +470,105 @@ class time_series_sstable_set : public sstable_set_impl {
     }
 };
 
+// Reader of a sstable together with its `max_position` (see `sstable::max_position`).
+struct reader_and_positions {
+    flat_mutation_reader reader;
+    position_in_partition max_position;
+
+    reader_and_positions(flat_mutation_reader r, position_in_partition max_pos)
+        : reader(std::move(r)), max_position(std::move(max_pos)) {}
+};
+
+// Queue of readers of sstables in a time_series_sstable_set,
+// returning readers in order of the sstables' min_position()s.
+//
+// Skips sstables that don't pass the supplied filter.
+//
+// The readers are created lazily on-demand using the supplied factory function.
+class min_position_reader_queue {
+    using container_t = time_series_sstable_set::container_t;
+    using value_t = container_t::value_type;
+
+    schema_ptr _schema;
+    lw_shared_ptr<const container_t> _sstables;
+
+    // Iterates over sstables in order of min_position().
+    container_t::const_iterator _it;
+    const container_t::const_iterator _end;
+
+    position_in_partition::tri_compare _cmp;
+
+    std::function<flat_mutation_reader(sstable&)> _create_reader;
```

Comment:
Will this be anything other than `sst->as_mutation_source()->create_reader()`in practice? Or is this just a way to bundle the mutation source params?

Reply:
In practice it will be:
```
    auto create_reader = [schema, permit, &pos, &slice, &pc, trace_state, fwd_sm] (sstable& sst) {
        return sst.read_row_flat(schema, permit, pos, slice, pc, trace_state, fwd_sm);
    };
```
1. it's a way to bind the parameters 2. it's a way to hide the details of reader creation from `min_position_reader_queue`, it just wants a factory function, it doesn't care how the readers are opened

Reply:
I think this should be converted to an sstable queue. Why even introduce readers here?

Reply:
Won't insist though.

Reply:
I'm now working on the tests and I think I found a second real implementation of a potential `min_position_reader_queue` interface: a test-specific implementation that returns readers from a set of readers created by the test.

So in the end it looks like we're going to do things the way you argued :) (and I'll probably put this stuff in mutation_reader module eventually)

But it also shows that we want this abstraction to return readers, not sstables.